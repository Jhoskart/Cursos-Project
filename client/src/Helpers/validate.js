function validate(data) {
    
    const validateName = (name) => {
      const regex = new RegExp(/^[0-9,$]*$/);
      return regex.test(name)
    }
      let errors = {};
      if (!data.name || data.name.length > 10)  errors.name = "No more than 10 characters!";
      if (validateName(data.name)) errors.name = "Name is required and must contain only letters!";
      if (!/.(gif|jpeg|jpg|png)$/i.test(data.image) && data.image !== "") { errors.image = "Must be a image url with format jpg, gif, png!";}
      if (!data.description || data.description.length < 20) errors.description = "Description must be at least 20 characters!";
      return errors;
    }
    
  export default validate;