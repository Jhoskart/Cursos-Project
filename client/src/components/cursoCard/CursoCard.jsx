import React from 'react'
import './cursoCard.css'
import CursoDetail from '../cursoDetail/CursoDetail'

export default function CursoCard({name, image, description, id}) {
    return (
        <div className='cards' >
            <div className='card card1'>
                <div className='container'>
                    <img src={image} alt="imagen del curso" />
                </div>
                <div className='details'>
                    <h3>{name}</h3>
                    <CursoDetail description={description} name={name} id={id} image={image}/>
                </div>
            </div>
        </div>
    )
}
