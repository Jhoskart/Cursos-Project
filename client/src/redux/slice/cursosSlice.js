import { createSlice } from '@reduxjs/toolkit';

export const cursosSlice = createSlice({
    name: 'cursos',
    initialState: {
        cursos: []
    },
    reducers: {
        getCursos: (state, {payload}) => {
            state.cursos = payload;
        }
    }
});

export const { getCursos } = cursosSlice.actions;