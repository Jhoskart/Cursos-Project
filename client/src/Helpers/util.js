import { render as rtlRender } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import {cursosSlice}from '../redux/slice/cursosSlice'
import { BrowserRouter as Router} from 'react-router-dom';

function reducer(ui, 
    { 
    preloadState,
    store = configureStore({ reducer: { cursos: cursosSlice.reducer },
    preloadState }),
    ...renderOptions
    } = {}
) {
    function Wrapper ({ children}) {
        return (
            <Provider store={store}>
                <Router>
                    {children}
                </Router>
            </Provider>
        );
    }
    return rtlRender(ui, { wrapper: Wrapper, ...renderOptions});
}

export * from '@testing-library/react'
export { reducer }

