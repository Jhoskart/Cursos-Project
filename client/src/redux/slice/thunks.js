import axios from 'axios';
import { getCursos } from './cursosSlice';
import Swal from 'sweetalert2';

const api = process.env.REACT_APP_API;

export const getCursosRequest = () => async dispatch => {
    try {
        const response = await axios.get(`${api}/cursos`);
        const data = await response.data;	
        dispatch(getCursos(data));
    } catch (error) {
        console.log(error);
    }
}

export const updateCursoRequest = (id, curso) => {
    return async () => {
        try {
            await axios.put(`${api}/cursos/${id}`, curso);
            Swal.fire({
                title: 'Curso actualizado',
                text: 'El curso se actualiz√≥ correctamente',
                icon: 'success',
                showConfirmButton: false,
                footer: '<a class="btn btn-primary" href="/">OK</a>'
            });
        } catch (error) {
            console.log(error);
        }
    }
}


export const postCursoRequest = (curso) => {
    return async () => {
        try {
            const response = await axios.post(`${api}/cursos`,{
            method: 'POST', 
            body: curso
        });	
            
            Swal.fire({
                title: 'Curso creado',
                text: 'El curso se ha creado correctamente',
                icon: 'success',
                showConfirmButton: false,
                footer: '<a class="btn btn-primary" href="/">OK</a>'
            });
            console.log(response);
        } catch (error) {
            Swal.fire({
                title: 'Error',
                text: `${error.response.data.message}`,
                icon: 'error',
                confirmButtonText: 'Ok'
            });
            console.log(error);
        }
    }
}

export const deleteCursoRequest = (id) => async dispatch => {
    try {
        const response = await axios.delete(`${api}/cursos/${id}`);
        Swal.fire({	
            title: 'Curso eliminado',
            text: 'El curso se ha eliminado correctamente',
            icon: 'success',
            showConfirmButton: false,
            footer: '<a class="btn btn-primary" href="/">OK</a>'
        });
        console.log(response);
    } catch (error) {
        Swal.fire({
            title: 'Error',
            text: `${error.response.data.message}`,
            icon: 'error',
            confirmButtonText: 'Ok'
        });
        console.log(error);
    }
}
