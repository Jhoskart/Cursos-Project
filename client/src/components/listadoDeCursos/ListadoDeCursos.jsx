import React, { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { getCursosRequest } from '../../redux/slice/thunks'
import CursoCard from '../cursoCard/CursoCard'
import CreateCurso from '../createCurso/CreateCurso'
import './listadoDeCursos.css'

export default function ListadoDeCursos() {
    const dispatch = useDispatch()
    const cursos = useSelector(state => state.cursos.cursos)
   
    useEffect(() => {
        dispatch(getCursosRequest())
    } , [dispatch])

    return (
        <div className="margin3">
            <h1>Listado de CursosðŸ‘½</h1>
            <div className="content">
                {cursos?.map(curso => (
                    <>
                        <CursoCard 
                            key={curso.id} 
                            name={curso.name} 
                            image={curso.image} 
                            description={curso.description}
                            id={curso.id}
                        />
                    </>
                ))}
            </div>
            <CreateCurso />  
        </div>
    )
}
