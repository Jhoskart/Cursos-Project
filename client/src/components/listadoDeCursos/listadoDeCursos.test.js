import { reducer, screen, cleanup } from '../../Helpers/util';
import { getCursos } from '../../redux/slice/cursosSlice'
import * as data from '../../db.json'
import thunk from 'redux-thunk';
import createMockStore from 'redux-mock-store'
import axios from 'axios'
axios.defaults.adapter = require('axios/lib/adapters/http');

import ListadoDeCursos from './ListadoDeCursos'
import CursoCard from '../cursoCard/CursoCard'

const mockStore = createMockStore([thunk]);

describe('First elements', () => {
    let title 
    beforeEach(() => {
        reducer(<ListadoDeCursos />)
        title = screen.queryByText('LarnU CursosðŸ‘½')
    });

    afterEach(() => {
        cleanup();
    });

    it('should render an h1 for the title', () => {
        expect(title).toBeInTheDocument()
    });

    it('should render a button to create a curso', () => {
        const button = screen.queryByText('Crear Curso')
        const component = reducer(<ListadoDeCursos />, button);
        expect(component.container).toMatchSnapshot()
    });

});

describe('<CursoCard />', () => {
    let cursoCard, state, store;
    let cursos = data.cursos[0];
    state = {
        cursos: []
    }
    store = mockStore(state)
    beforeEach(() => {
        cursoCard = (cursos) =>
            reducer(
                <CursoCard
                    name={cursos.name}
                    description={cursos.description}
                    image={cursos.image}
                    id={cursos.id}
                />
            )
    }),
    
    afterEach(() => {
        jest.restoreAllMocks();
    });

    it('should render correctly', () => {
        const { container } = cursoCard(cursos)
        expect(container).toMatchSnapshot();
    });

    it('should render a div container', () => {
        const { container } = cursoCard(cursos)
        expect(container.querySelector('div')).toHaveClass('cards');
    });

    it('should render an h1 for the title', () => {
        const { getByText } = cursoCard(cursos)
        expect(getByText('Curso de NodeJS')).toBeInTheDocument()
    });

    it('should render an img tag with altText', () => {
        const { getByRole } = cursoCard(cursos)
        expect(getByRole("img", { name: "imagen del curso" })).toBeInTheDocument()
    });
});

describe('GetCursos action thunk', () => {
    it('dispatch getCursos action thunk ', async () => {
        const store = mockStore()
        await store.dispatch(getCursos(data.cursos))
        const actions = store.getActions();
        expect(actions[0]).toEqual(getCursos(data.cursos));
    });
});

