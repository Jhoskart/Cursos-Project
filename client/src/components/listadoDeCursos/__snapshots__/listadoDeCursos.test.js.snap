// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CursoCard /> should render correctly 1`] = `
<div>
  <div
    class="cards"
  >
    <div
      class="card card1"
    >
      <div
        class="container"
      >
        <img
          alt="imagen del curso"
          src="http://www.jorgehernandezramirez.com/wp-content/uploads/2017/07/nodelogo.png"
        />
      </div>
      <div
        class="details"
      >
        <h3>
          Curso de NodeJS
        </h3>
        <div>
          <button
            class="btn btn-light"
            type="button"
          >
            Mas informacion aqui
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`First elements should render a button to create a curso 1`] = `
<div>
  <div
    class="margin3"
  >
    <h1>
      LarnU CursosðŸ‘½
    </h1>
    <div
      class="content"
    />
    <div
      class="margin2"
    >
      <button
        class="tamaÃ±o btn btn-primary"
        type="button"
      >
        Crear Curso
      </button>
    </div>
  </div>
</div>
`;
